src/ovs_events.c

```diff
@@ -0,0 +1,629 @@
+       iface = iface->next)
+    return (strcmp(ifname, iface->name) == 0);
+
+  return 0;
+}
+
+/* Get OVS DB select parameter request based on rfc7047,
+ * "Transact" & "Select" section
+ */
+static char *ovs_events_get_select_params() {
+  int ret = 0;
+  size_t buff_size = 0;
+  size_t offset = 0;
+  char *buff = NULL;
+  char *new_buff = NULL;
+  const char params_fmt[] = "[\"Open_vSwitch\"%s]";
+  const char option_fmt[] = ",{\"op\":\"select\",\"table\":\"Interface\","
+                            "\"where\":[[\"name\",\"==\",\"%s\"]],"
+                            "\"columns\":[\"link_state\",\"external_ids\","
+                            "\"name\",\"_uuid\"]}";
+  const char default_opt[] = ",{\"op\":\"select\",\"table\":\"Interface\","
+                             "\"where\":[],\"columns\":[\"link_state\","
+                             "\"external_ids\",\"name\",\"_uuid\"]}";
+  /* setup OVS DB interface condition */
+  for (ovs_events_iface_list_t *iface = ovs_events_ctx.config.ifaces; iface;
+       iface = iface->next, offset += ret) {
+    /* allocate new buffer (format size + ifname len is good enough) */
+    buff_size += (sizeof(option_fmt) + strlen(iface->name));
+    new_buff = realloc(buff, buff_size);
+    if (new_buff == NULL)
+      goto failure;
+    buff = new_buff;
+    ret = ssnprintf(buff + offset, buff_size, option_fmt, iface->name);
+    if (ret < 0)
+      goto failure;
+  }
+  /* if no interfaces are configured, use default params */
+  if (buff == NULL) {
+    buff = strdup(default_opt);
+    offset = strlen(default_opt);
+  }
+
+  /* allocate memory for OVS DB select params */
+  buff_size = offset + sizeof(params_fmt);
+  new_buff = malloc(buff_size);
+  if (new_buff == NULL)
+    goto failure;
+
+  /* create OVS DB select params */
+  if (ssnprintf(new_buff, buff_size, params_fmt, buff) < 0)
+    goto failure;
+
+  sfree(buff);
+  return new_buff;
+
+failure:
+  sfree(new_buff);
+  sfree(buff);
+  return NULL;
+}
+
+/* Release memory allocated for configuration data */
+static void ovs_events_config_free() {
+  ovs_events_iface_list_t *del_iface = NULL;
+  sfree(ovs_events_ctx.ovs_db_select_params);
+  while (ovs_events_ctx.config.ifaces) {
+    del_iface = ovs_events_ctx.config.ifaces;
+    ovs_events_ctx.config.ifaces = ovs_events_ctx.config.ifaces->next;
+    sfree(del_iface);
+  }
+}
+
+/* Parse plugin configuration file and store the config
+ * in allocated memory. Returns negative value in case of error.
+ */
+static int ovs_events_plugin_config(oconfig_item_t *ci) {
+  ovs_events_iface_list_t *new_iface;
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+    if (strcasecmp("SendNotification", child->key) == 0) {
+      if (cf_util_get_boolean(child, &ovs_events_ctx.config.send_notification) <
+          0)
+        OVS_EVENTS_CONFIG_ERROR(child->key);
+    } else if (strcasecmp("OvsDbAddress", child->key) == 0) {
+      if (child->values_num < 1) {
```

Comment:
Please use `cf_util_get_string_buffer()` here, which will do all necessary checks and error reporting:

```c
cf_util_get_string_buffer(child, ovs_events_ctx.config.ovs_db_node,
    sizeof(ovs_events_ctx.config.ovs_db_node));
```